src/mono/browser/runtime/cancelable-promise.ts

```diff
@@ -32,17 +38,121 @@ export function wrap_as_cancelable<T>(inner: Promise<T>): ControllablePromise<T>
-    holder.isCanceled = true;
-    promise_control.reject(new Error("OperationCanceledException"));
+// NOTE: layout has to match PromiseHolderState in JSHostImplementation.Types.cs
+const enum PromiseHolderState {
+    IsResolving = 0,
 }
 
+const promise_holder_symbol = Symbol.for("wasm promise_holder");
+
+export class PromiseHolder extends ManagedObject {
+    public isResolved = false;
+    public isPosted = false;
+    public data: any = null;
+    public reason: any = null;
+    public constructor(public promise: Promise<any>,
+        private gc_handle: GCHandle,
+        private promiseHolderPtr: number, // could be null for GCV_handle 
+        private res_converter?: MarshalerToCs) {
+        super();
+    }
+
+    // returns true if the promise is being canceled by managed code
+    setIsResolving(): boolean {
+        if (!WasmEnableThreads || this.promiseHolderPtr === 0) {
+            return false;
+        }
+        forceThreadMemoryViewRefresh();
+        if (compareExchangeI32(this.promiseHolderPtr + PromiseHolderState.IsResolving, 1, 0) === 0) {
+            return false;
+        }
+        return true;
+    }
+
+    resolve(data: any) {
+        if (this.isResolved) {
+            return;
+        }
+        if (WasmEnableThreads && this.setIsResolving()) {
+            // we know that cancelation is in flight
+            // because we need to keep the GCHandle alive until until the cancelation arrives
+            // we skip the this resolve and let the cancelation to reject the Task
+            // we store the original data and use it later
+            this.data = data;
+            return;
+        }
+        this.isResolved = true;
+        this.complete_task(data, null);
+    }
+
+    reject(reason: any) {
+        const isCancelation = reason && reason[promise_holder_symbol] === this;
+        if (this.isResolved && !isCancelation) {
+            return;
+        }
+        if (WasmEnableThreads && !isCancelation && this.setIsResolving()) {
+            // we know that cancelation is in flight
+            // because we need to keep the GCHandle alive until until the cancelation arrives
+            // we skip the this reject and let the cancelation to reject the Task
+            // we store the original reason and use it later
+            this.reason = reason;
+            return;
+        }
+        this.isResolved = true;
+        this.complete_task(null, reason);
+    }
+
+    cancel() {
+        if (this.isResolved) {
+            return;
+        }
+        this.isResolved = true;
+
+        const promise = this.promise;
+        loaderHelpers.assertIsControllablePromise(promise);
+        const promise_control = loaderHelpers.getPromiseController(promise);
+
+        const reason = this.reason || new Error("OperationCanceledException") as any;
+        reason[promise_holder_symbol] = this;
+
+        if (promise_control.isDone) {
+            this.setIsResolving();
```

Comment:
should we be checking the return value of this and only calling complete_task conditionally? it seems like if we somehow end up calling complete_task twice, it will assert and take the app down